tensorflow/python/keras/layers/preprocessing/preprocessing_stage_functional.py

```diff
@@ -0,0 +1,141 @@
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# ==============================================================================
+"""Preprocessing stage."""
+# pylint: disable=g-classes-have-attributes
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+
+import numpy as np
+
+from tensorflow.python.data.ops import dataset_ops
+from tensorflow.python.framework import ops
+from tensorflow.python.keras.engine import base_preprocessing_layer
+from tensorflow.python.keras.engine import functional
+from tensorflow.python.keras.utils import tf_utils
+from tensorflow.python.util import nest
+
+
+class PreprocessingStageFunctional(base_preprocessing_layer.PreprocessingLayer,
+                                   functional.Functional):
+  """A functional preprocessing stage.
+
+  This preprocessing stage wraps a graph of preprocessing layers into a
+  Functional-like object that enables you to `adapt()` the whole graph via
+  a single `adapt()` call on the preprocessing stage.
+
+  Arguments:
+    inputs: List of input tensors (must be created via `tf.keras.Input()`).
+    outputs: List of outputs tensors.
+    name: String, optional. Name of the model.
+    trainable: Boolean, whether the model's variables should be trainable.
+  """
+
+  def adapt(self, data, reset_state=True):
+    """Adapt the state of the layers of the preprocessing stage to the data.
+
+    Arguments:
+      data: A batched Dataset object, or a NumPy array, or an EagerTensor.
+        Data to be iterated over to adapt the state of the layers in this
+        preprocessing stage.
+      reset_state: Whether this call to `adapt` should reset the state of
+        the layers in this preprocessing stage.
+    """
+    if not isinstance(data, dataset_ops.Dataset):
+      data = self._flatten_to_reference_inputs(data)
+      if any([not isinstance(datum, (np.ndarray, ops.EagerTensor))
+              for datum in data]):
+        raise ValueError(
+            '`adapt()` requires a batched Dataset, a list of EagerTensors '
+            'or Numpy arrays as input, got {}'.format(type(data)))
+
+    if isinstance(data, dataset_ops.Dataset):
+      # Validate the datasets to try and ensure we haven't been passed one with
+      # infinite size. That would cause an infinite loop here.
+      if tf_utils.dataset_is_infinite(data):
+        raise ValueError(
+            'The dataset passed to `adapt()` has an infinite number of '
+            'elements. Please use dataset.take(...) to make the number '
+            'of elements finite.')
+      # Unzip dataset object to a list of single input dataset.
+      data = _unzip_dataset(data)
```

Comment:
This will go out of memory instantly. You should iterate over the dataset one batch at a time (by calling `adapt(ds)` on individual layers)